test/copy.cpp

```diff
@@ -0,0 +1,170 @@
+	const char * const sourceUriString = "https://example.com";
+
+	int result = uriParseSingleUriExA(&state, sourceUriString, sourceUriString + strlen(sourceUriString), NULL);
+	ASSERT_EQ(URI_SUCCESS, result);
+
+	UriUriA destUri;
+	ASSERT_EQ(URI_ERROR_NULL, uriCopyUriA(NULL, &sourceUri));
+
+	uriFreeUriMembersA(&sourceUri);
+}
+
+TEST(CopyUriSuite, ErrorIncompleteMemoryManager) {
+	UriParserStateA state;
+	UriUriA sourceUri;
+	state.uri = &sourceUri;
+	const char * const sourceUriString = "https://example.com";
+	UriMemoryManager memory;
+
+	int result = uriParseSingleUriExA(&state, sourceUriString, sourceUriString + strlen(sourceUriString), NULL);
+	ASSERT_EQ(URI_SUCCESS, result);
+
+	UriUriA destUri;
+	ASSERT_EQ(URI_ERROR_MEMORY_MANAGER_INCOMPLETE, uriCopyUriMmA(NULL, &sourceUri, &memory));
+
+	uriFreeUriMembersA(&sourceUri);
+}
+
+TEST(CopyUriSuite, Success) {
+	uriUriA * destUri = testCopyUri("https://somehost.com");
+
+	ASSERT_TRUE(destUri->hostData->ipv4 == NULL);
+	ASSERT_TRUE(destUri->hostData->ipv6 == NULL);
+	ASSERT_EQ(0, strncmp(destUri->hostData->ipFuture.first, 7);
+	ASSERT_TRUE(destUri->hostData->ipFuture.afterLast == NULL);
+
+	uriFreeUriMembersA(&destUri);
+}
+
+TEST(CopyUriSuite, SuccessCompleteUri) {
+	uriUriA * destUri = testCopyUri("https://user:pass@somehost.com:80/path?query#frag");
+
+	uriFreeUriMembersA(&destUri);
+}
+
+TEST(CopyUriSuite, SuccessRelativeReference) {
+	uriUriA * destUri = testCopyUri("/foo/bar/baz");
+
+	uriFreeUriMembersA(&destUri);
+}
+
+TEST(CopyUriSuite, SucessEmail) {
+	uriUriA * destUri = testCopyUri("mailto:fred@example.com");
+
+	uriFreeUriMembersA(&destUri);
+}
+
+TEST(CopyUriSuite, SuccessIpv4) {
+	uriUriA * destUri = testCopyUri("http://192.168.0.1/");
+
+	ASSERT_EQ(192, destUri->hostData->ipv4.data[0]);
+	ASSERT_EQ(168, destUri->hostData->ipv4.data[1]);
+	ASSERT_EQ(0, destUri->hostData->ipv4.data[2]);
+	ASSERT_EQ(1, destUri->hostData->ipv4.data[3]);
+
+	ASSERT_TRUE(destUri->hostData->ipv6 == NULL);
+	ASSERT_TRUE(destUri->hostData->ipFuture.first == NULL);
+	ASSERT_TRUE(destUri->hostData->ipFuture.afterLast == NULL);
+
+	uriFreeUriMembersA(&destUri);
+}
+
+TEST(CopyUriSuite, SuccessIpv6) {
+	uriUriA * destUri = testCopyUri("https://[2001:0db8:0001:0000:0000:0ab9:C0A8:0102]");
+
+	ASSERT_EQ(32, destUri->hostData->ipv6[0]);
+	ASSERT_EQ(1, destUri->hostData->ipv6[1]);
+	ASSERT_EQ(13, destUri->hostData->ipv6[2]);
+	ASSERT_EQ(184, destUri->hostData->ipv6[3]);
+	ASSERT_EQ(0, destUri->hostData->ipv6[4]);
+	ASSERT_EQ(1, destUri->hostData->ipv6[5]);
+	ASSERT_EQ(0, destUri->hostData->ipv6[6]);
+	ASSERT_EQ(0, destUri->hostData->ipv6[7]);
+	ASSERT_EQ(0, destUri->hostData->ipv6[8]);
+	ASSERT_EQ(0, destUri->hostData->ipv6[9]);
+	ASSERT_EQ(10, destUri->hostData->ipv6[10]);
+	ASSERT_EQ(185, destUri->hostData->ipv6[11]);
+	ASSERT_EQ(192, destUri->hostData->ipv6[12]);
+	ASSERT_EQ(168, destUri->hostData->ipv6[13]);
+	ASSERT_EQ(1, destUri->hostData->ipv6[14]);
+	ASSERT_EQ(2, destUri->hostData->ipv6[15]);
```

Comment:
@kocsismate based on typo "ipv6" instead of "ip6" I am now assuming that you only pushed for backup and not for review… please ping back when ready for review :pray:

Reply:
Sorry, I wanted to try out my changes in CI, but I feel asleep immediately, and I didn't have time to fix my changes. You are right, I should have tried them out locally, because there were a lot of issues with it. Now, the tests succeed locally, but I'll have another look later, so please do not review it yet.